src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_fsm.cpp

```diff
@@ -0,0 +1,172 @@
+      best_pass_and_score_so_far(
+          PassWithRating{.pass = Pass(Point(), Point(), 0), .rating = 0}),
+      time_since_commit_stage_start(Duration::fromSeconds(0)),
+      min_pass_score_threshold(0)
+{
+}
+
+void ShootOrPassPlayFSM::updateOffensivePositioningTactics(
+    std::vector<EighteenZoneId> ranked_zones, PassEvaluation<EighteenZoneId> pass_eval,
+    unsigned int num_tactics)
+{
+    // These two tactics will set robots to roam around the field, trying to put
+    // themselves into a good position to receive a pass
+    if (num_tactics != offensive_positioning_tactics.size())
+    {
+        offensive_positioning_tactics =
+            std::vector<std::shared_ptr<MoveTactic>>(num_tactics);
+        std::generate(offensive_positioning_tactics.begin(),
+                      offensive_positioning_tactics.end(),
+                      []() { return std::make_shared<MoveTactic>(false); });
+    }
+
+    for (unsigned int i = 0; i < offensive_positioning_tactics.size(); i++)
+    {
+        auto pass1 = pass_eval.getBestPassInZones({ranked_zones[i]}).pass;
+
+        offensive_positioning_tactics[i]->updateControlParams(
+            pass1.receiverPoint(), pass1.receiverOrientation(), 0.0,
+            MaxAllowedSpeedMode::PHYSICAL_LIMIT);
+    }
+}
+
+void ShootOrPassPlayFSM::lookForPass(const Update& event)
+{
+    PriorityTacticVector ret_tactics = {{attacker_tactic}, {}};
+
+    // only look for pass if there are more than 1 robots
+    if (event.common.num_tactics > 1)
+    {
+        auto pitch_division =
+            std::make_shared<const EighteenZonePitchDivision>(event.common.world.field());
+
+        auto pass_eval    = pass_generator.generatePassEvaluation(event.common.world);
+        auto ranked_zones = pass_eval.rankZonesForReceiving(
+            event.common.world, event.common.world.ball().position());
+
+        best_pass_and_score_so_far = pass_eval.getBestPassOnField();
+
+
+        // Wait for a good pass by starting out only looking for "perfect" passes
+        // (with a score of 1) and decreasing this threshold over time
+        // This boolean indicates if we're ready to perform a pass
+        double abs_min_pass_score =
+            play_config->getShootOrPassPlayConfig()->getAbsMinPassScore()->value();
+        double pass_score_ramp_down_duration = play_config->getShootOrPassPlayConfig()
+                                                   ->getPassScoreRampDownDuration()
+                                                   ->value();
+        pass_eval = pass_generator.generatePassEvaluation(event.common.world);
+        best_pass_and_score_so_far = pass_eval.getBestPassOnField();
+
+        // update the best pass in the attacker tactic
+        attacker_tactic->updateControlParams(best_pass_and_score_so_far.pass, false);
+
+        // If we've assigned a robot as the passer in the PassGenerator, we
+        // lower our threshold based on how long the PassGenerator as been
```

Comment:
nit: how long the PassGenerator as been running -> how long the PassGenerator has been running

Reply:
fixed